### PR TITLE
Prometheus: Improve expanding ruleIs with identifier label checking

### LINKED ISSUE #87659: Prometheus Data Source - Expand Rules: When multiple recording rules match a metric-name the first match isn't always correct
### What happened?

FWIW - I'm writing this up after discussing a customer's experience with @itsmylife

I was looking at an Mimir alert rule definition, and chose to "Explore" the query. The Explore GUI detected that the query was referring to a recorded series and suggested I could "expand rules", which I did. The results weren't what I expected because I have more than one recording rule that writes into the metric name, each made unique by additional prometheus labels on the recorded series.

### What did you expect to happen?

I expected the "expand rules" function to find the rule that populated the series I was querying or to give me some kind of warning / error that it found multiple matches and might not work for me query.

I also expected the result of the "expand rules" action to be a promql query that no longer contained the `uuid=...` label matcher, because that label was added as the result of the recording rule and didn't exist in the raw metrics (i.e. I can't run the expression from the rule AND include the "uuid" label matcher, that uuid label won't exist on the raw metrics, only the recorded series)

https://github.com/grafana/grafana/assets/5599664/90acfd75-0081-4d52-b6f7-d41d8b876773

### Did this work before?

No, I don't think this worked before. I chose to write it up as a bug rather than a feature request, because I thought it would be easier to understand the issue this way.

### How do we reproduce it?

1. Produce multiple recording rules that write to the same metric name (ex: `recorded_series`) , where each adds labels to the rule object (ex: `uuid=<unique value per rule>`)
2. Explore a promql query like `recorded_series{uuid="foo"}` and click "expand rules" 
3. Confirm that the expanded query comes from the first recording rule that matches metric name `recorded_series` rather than from the one that had `rule.Labels[uuid] = "foo"`

### Is the bug inside a dashboard panel?

no. It's in the Explore GUI

### Environment (with versions)?

Grafana: 11.1.0
OS: Cloud
Browser: chrome

### Grafana platform?

I use Grafana Cloud

### Datasource(s)?

prometheus

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

In Prometheus we have recording rule groups. https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/ Each group can have a rule with a name used in another group. In general we use labels to distinguish them.

And in Grafana we have a feature to expand that recording rule. That is just unraveling the query that the recording rule holds. But in the cases above there is no way to know which query the user wants to see. Even providing identifier label won't work because there was no such feature to look for an identifier label. 

So this PR introduces two new things on top of the existing "expand rule" feature:
1. Expand the Rule based on the identifier label/value
  a. Identifier label must be a label that has `uuid` keyword in it. 
3. Show a warning for cases that the same rule name appears in two or more different rule groups

Please refer [this issue](https://github.com/grafana/grafana/issues/87659) to more in-depth explanation.

**Why do we need this feature?**

To be able to expand the recording rules based on the identifier labels

**Who is this feature for?**

Prometheus users who have the same recording rule name in different rule groups. 

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/87659

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-prometheus/src/query_hints.ts [hunk 1]
```diff
@@ -230,6 +210,125 @@ export function getInitHints(datasource: PrometheusDatasource): QueryHint[] {
...
+        ruleName,
+        mapping[ruleName]
+      );
+
+      // No identifier detected add warning
+      if (idx === -1) {
+        hints.push({
+          type: 'EXPAND_RULES_WARNING',
+          label: 'Query contains a recording rule. To be able to expand it please add an identifier label/value.',
+        });
```
Comment (itsmylife) [inline packages/grafana-prometheus/src/query_hints.ts:232]: @catherineymgui @ellisda Please comment on this text.

Comment (ellisda) [reply packages/grafana-prometheus/src/query_hints.ts:232]:
```suggestion
        hints.push({
          type: 'EXPAND_RULES_WARNING',
          label: 'Query contains recording rules and multiple matches found. To be able to expand it please add an identifier label/value.',
        });
```

The previous message is pretty terse, just "Query contains recording rules" so I'm not sure how much additional verbosity fits... :thinking: 

I think the key is to say that we matched or found "multiple recording rules"

I think your last sentence is good.

Comment (catherineymgui) [reply packages/grafana-prometheus/src/query_hints.ts:232]: I think it's good to make it as short as possible but to the point. How about this? 

> We found multiple recording rules in this query. To expand the recording rule, add an identifier label/value.

I removed the word please and made it active voice [as per these guidelines](https://docs.google.com/document/d/1Vx6EHVDTOpz_sUmM5Mtkx3XTMEj0bHen6mmStnp1IR4/edit)

Comment (ellisda) [reply packages/grafana-prometheus/src/query_hints.ts:232]: I think that works. Keeping with active voice, I think "We found multiple recording rules that match in this query" might be better. The "match" is the key here. But I like your current suggestion as well.

FWIW there is a difference between:
1. finding multiple recording rules in the query `metric1 - metric2` - where both match a recording rule
2. finding a single recording rule in the query `metric` that matches with multiple recording rule definitions that each record into a metric named `metric`

File: packages/grafana-prometheus/src/query_hints.ts [hunk 2]
```diff
@@ -230,6 +210,125 @@ export function getInitHints(datasource: PrometheusDatasource): QueryHint[] {
...
+  if (queryMetricLabels.length === 0) {
+    return { idx: -1, identifier: '', identifierValue: '', expandedQuery: '' };
+  }
+
+  let uuidLabel = '';
+  let uuidLabelValue = '';
+  let uuidLabelIdx = -1;
+
+  queryMetricLabels.forEach((qml) => {
+    if (uuidLabelIdx === -1 && qml.label.search('uuid') !== -1) {
```
Comment (ellisda) [inline packages/grafana-prometheus/src/query_hints.ts:290]: question: is it worth trying to generalize this and work with other labels that don't include "uuid" in the label name?

For Grafana SLO use cases, the "uuid" is sufficient. Maybe this is an easier place to start and we can generalize it later if we care to ?

**Review by ellisda — APPROVED** [review]
LGTM :partying_face: 

I'm not a typescript expert, but I did pull the PR build and validate the primary use case for Grafana SLO :balloon: 

https://github.com/user-attachments/assets/d363cdc1-0f51-4321-b17f-b55e7fb09bd1

File: packages/grafana-prometheus/src/query_hints.ts (hunk 2 above)
Comment (itsmylife) [reply packages/grafana-prometheus/src/query_hints.ts:291]: Not in this PR but maybe in a separate task we should definitely take a look at that. Do you have a suggestion how we would detect the identifier label?

**Comment by bohandley** [comment]
It would be helpful to have a description of what it takes to set up the rules scenario to test this PR :) Or an example of a rule file that fits this edge case.

**Review by bohandley — COMMENTED** [review]
@itsmylife @ellisda 

I have added this to my rules file and tested it. A rule with two expressions where a user should use a label, `uuid` to identify the different expressions. Is this a correct test scenario?

```
groups:
  - name: recording_rules
    interval: 5s
    rules:
      - record: a_rule:with_two_expr
        expr: histogram_quantile(0.9, rate(go_gc_heap_allocs_by_size_bytes_bucket[5m]))
        labels:
          uuid: 123
      - record: a_rule:with_two_expr
        expr: go_goroutines
        labels:
          uuid: 456
      - record: a_rule:with_one_expr
        expr: rate(go_goroutines{}[5m])
```

It looks good for the code editor. The message pops up for the rule with two expression about adding a label.

Screenshot 2024-07-17 at 9 31 11 AM

But for the query editor, no warning pops up. Can we add the warning to the query editor as well?

Screenshot 2024-07-17 at 9 31 53 AM

**Comment by ellisda** [comment]
> Is this a correct test scenario?

@bohandley , yes that is a correct test scenario where `a_rule:with_two_expr` needs a `{uuid="123"}` to know which rule to use in expand.

TIL that "expand rules" also appears in builder mode (I use code view much more often)

image

**Review by bohandley — APPROVED** [review]
Upon further inspection of the query builder, we are only showing the warning hints on init and hints that have an action (fix field) so that the button populates.

https://github.com/grafana/grafana/blob/1404249ff151839bb91bc2266e058fd8ffee9e97/packages/grafana-prometheus/src/querybuilder/shared/QueryBuilderHints.tsx#L36

I think we should create an issue to show warning hints (no fix) in the query builder but this shouldn't stop this PR. This would be a good first PR for developers.

I'm approving because the expand rules button will show if the user selects the appropriate label, `uuid` in the query builder. It is confusing though to select a metric that appears to be a recording rule and not see the `expand rules` pop up until a label is selected. The warning would be nice improvement for UX but we can add that to the issue.

Screenshot 2024-07-17 at 9 54 02 AM

🚀

File: packages/grafana-prometheus/src/query_hints.ts [hunk 3]
```diff
@@ -230,6 +210,126 @@ export function getInitHints(datasource: PrometheusDatasource): QueryHint[] {
...
+        ruleName,
+        mapping[ruleName]
+      );
+
+      // No identifier detected add warning
+      if (idx === -1) {
+        hints.push({
+          type: 'EXPAND_RULES_WARNING',
+          label:
+            'We found multiple recording rules that match in this query. To expand the recording rule, add an identifier label/value.',
```
Comment (bohandley) [inline packages/grafana-prometheus/src/query_hints.ts:232]: When a hint does not have a fix or action it won't show up in the query builder. This warning is only shown in the code editor.

https://github.com/grafana/grafana/blob/1404249ff151839bb91bc2266e058fd8ffee9e97/packages/grafana-prometheus/src/querybuilder/shared/QueryBuilderHints.tsx#L36

Comment (itsmylife) [reply packages/grafana-prometheus/src/query_hints.ts:232]: In the future we will have an action for that warning. But for now, that's how it is. And there is a line above the one you linked, it says `For now show only actionable hints` so I think this is fine for now.

**Pull request merged**